File: VB/FriendlyUrlSample.Module.Web/Controllers/CustomLinkController.vb
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.Web
Imports DevExpress.Persistent.Base
Imports FriendlyUrlSample.Module.FriendlyUrlSample.Module.BusinessObjects

Namespace FriendlyUrlSample.[Module].Web.Controllers
    Public Class CustomLinkController
        Inherits ViewController

        Private goToJohnNilsenAction As SimpleAction

        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            Dim johnNilsen As Contact = ObjectSpace.FindObject(Of Contact)(CriteriaOperator.Parse("LastName == 'Nilsen'"))

            If johnNilsen IsNot Nothing Then
                goToJohnNilsenAction.Active.RemoveItem("JohnNilsenIsNotExist")
                Dim viewShortcut As ViewShortcut = New ViewShortcut(Application.GetDetailViewId(GetType(Contact)), ObjectSpace.GetKeyValueAsString(johnNilsen))
                Dim url As String = (CType(Application, WebApplication)).ViewUrlManager.GetUrl(viewShortcut)
                goToJohnNilsenAction.SetClientScript($"window.open('{url}', '_blank')", False)
            Else
                goToJohnNilsenAction.Active("JohnNilsenIsNotExist") = False
            End If
        End Sub

        Public Sub New()
            goToJohnNilsenAction = New SimpleAction(Me, "GoToJohnNilsen", PredefinedCategory.View)
        End Sub

        Private Sub InitializeComponent()
        End Sub
    End Class
End Namespace

File: CS/FriendlyUrlSample.Module.Web/Controllers/CustomLinkController.cs
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.Web;
using DevExpress.Persistent.Base;
using FriendlyUrlSample.Module.BusinessObjects;

namespace FriendlyUrlSample.Module.Web.Controllers {
    public class CustomLinkController : ViewController {
        private SimpleAction goToJohnNilsenAction;

        protected override void OnActivated() {
            base.OnActivated();
            Contact johnNilsen = ObjectSpace.FindObject<Contact>(CriteriaOperator.Parse("LastName == 'Nilsen'"));
            if(johnNilsen != null) {
                goToJohnNilsenAction.Active.RemoveItem("JohnNilsenIsNotExist");
                ViewShortcut viewShortcut = new ViewShortcut(Application.GetDetailViewId(typeof(Contact)), ObjectSpace.GetKeyValueAsString(johnNilsen));
                string url = ((WebApplication)Application).ViewUrlManager.GetUrl(viewShortcut);
                goToJohnNilsenAction.SetClientScript($"window.open('{url}', '_blank')", false);
            }
            else {
                goToJohnNilsenAction.Active["JohnNilsenIsNotExist"] = false;
            }
        }
        public CustomLinkController() {
            goToJohnNilsenAction = new SimpleAction(this, "GoToJohnNilsen", PredefinedCategory.View);
        }
    }
}
